http-cache-tower/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+    inner: S,
+    cache: HttpCache<T>,
+}
+
+impl<S, T> Cache<S, T>
+where
+    T: Clone + CacheManager,
+{
+    /// Create a new [`Cache`].
+    pub fn new(inner: S, cache: HttpCache<T>) -> Self {
+        Self { inner, cache }
+    }
+
+    /// Returns a new [`Layer`] that wraps services with a `Cache` middleware.
+    pub fn layer(&self) -> CacheLayer<T> {
+        CacheLayer::new(self.cache.clone())
+    }
+}
+
+/// [`Layer`] with a [`Service`] to cache responses.
+#[derive(Clone, Debug)]
+pub struct CacheLayer<T: CacheManager> {
+    cache: HttpCache<T>,
+}
+
+impl<T> CacheLayer<T>
+where
+    T: CacheManager,
+{
+    /// Create a new [`CacheLayer`].
+    pub fn new(cache: HttpCache<T>) -> Self {
+        Self { cache }
+    }
+}
+
+impl<S, T> Layer<S> for CacheLayer<T>
+where
+    S: Clone,
+    T: Clone + CacheManager,
+{
+    type Service = Cache<S, T>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Cache { inner, cache: self.cache.clone() }
+    }
+}
+
+impl<ReqBody, ResBody, S, T> Service<Request<ReqBody>> for Cache<S, T>
+where
+    S: Clone + Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ReqBody: Body + Clone,
+    T: Clone + CacheManager,
+{
+    type Response = Response<ResBody>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S, ReqBody, T>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<std::result::Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let service = self.inner.clone();
+        let mut service = mem::replace(&mut self.inner, service);
+        ResponseFuture {
+            req: req.clone(),
+            future: Either::Left(service.call(req)),
+            service,
+            cache: self.cache.clone(),
+            state: ResponseFutureState::Init,
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`Cache`].
+    #[derive(Debug)]
+    pub struct ResponseFuture<S, B, T>
+    where
+        S: Service<Request<B>>,
+        T: CacheManager,
+    {
+        #[pin]
+        future: Either<S::Future, Oneshot<S, Request<B>>>,
+        service: S,
+        req: Request<B>,
+        cache: HttpCache<T>,
+        state: ResponseFutureState<F>,
+    }
+}
+
+impl<S, ReqBody, ResBody, T> Future for ResponseFuture<S, ReqBody, T>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone,
+    ReqBody: Body + Clone,
+    T: CacheManager + Clone,
+{
+    type Output = std::result::Result<Response<ResBody>, S::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+        let mut curr_state = this.state;
+        match this.state.as_mut().project() {
+            ResponseFutureProj::Init => {
+                let request_parts = this.req.clone().into_parts().0;
+                let fut = this.cache.before_request(&request_parts);
```

Comment:
@06chaynes, I took this line from the commented-out code in your branch, but there's no `before_request` method on `cache`. Assuming it's because the previous branch was out of date, what should be called here instead?

If you have an async version of the code, I can probably translate it to the hand-rolled future

Reply:
@SebRollen yeah in that branch I was experimenting with a different implementation that I decided (at least for now) not to go with, though I did bring over a handful of changes. 

Here is the implementation for Reqwest that has the logic as it currently is https://github.com/06chaynes/http-cache/blob/develop/http-cache-reqwest/src/lib.rs#L212

First the `Middleware` trait implementation is constructed, then the request is checked to determine if it's cacheable. If it isnt cacheable we just run the `run_no_cache` method of the `HttpCache` struct, handle the response normally, update the two cache headers, and pass it through. If it is cacheable we run the `run` method of the `HttpCache` struct and pass through the result.

Reply:
pushed a commit to update the core and other crates http and http-cache-semantics versions